### PR TITLE
changes from supermuc

### PR DESCRIPTION
**Pull request by paulstapor**
This branch contains a fixed version of the minibatch code concerning
 - the rescue interceptor
 - the line search implementation

### CODE REVIEW

File: include/parpeoptimization/minibatchOptimization.h [hunk 1]
```diff
@@ -393,15 +393,15 @@ class MinibatchOptimizer {
                     learningRateUpdater->increaseLearningRate();
 
                     // Overwrite old parameters and old gradient, since they won't be needed any more
-                    oldGradient = gradient;
-                    oldParameters = parameters;
+                    for (int i = 0; i < parameters.size(); ++i) {
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: Why the change?

File: include/parpeoptimization/minibatchOptimization.h [hunk 2]
```diff
@@ -509,51 +509,94 @@ class MinibatchOptimizer {
                                                OptimizationReporter *reporter) {
         int maxSubsequentFails = 10;
         bool finalFail = false;
+        bool initialFail = false;
         cost = NAN;
         FunctionEvaluationStatus status = functionEvaluationFailure;
+        std::vector<double> parDifference;
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]:
```suggestion
        std::vector<double> parDifference(parameters.size());
```
and remove .resize()

File: include/parpeoptimization/minibatchOptimization.h [hunk 3]
```diff
@@ -509,51 +509,94 @@ class MinibatchOptimizer {
...
             reporter->logger->setPrefix(logger->getPrefix());
         }
+        
+        // debug output
+        
+        for (int i = 0; i < parameters.size(); ++i)
+            firstDifference[i] = parameters[i] - oldParameters[i];
+        std::stringstream first_ss;
+        first_ss << " Interceptor is active! Former step size: " << getVectorNorm(firstDifference) << std::endl;
+        if (logger)
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: The 4 lines above should be moved to the if body

File: include/parpeoptimization/minibatchOptimization.h [hunk 4]
```diff
@@ -509,51 +509,94 @@ class MinibatchOptimizer {
...
 
         // Cost function evaluation failed: We need to intercept
         while (status == functionEvaluationFailure) {
+            
             // If the objective function evaluation failed, we want to undo the step
             ++subsequentFails;
             parameterUpdater->undoLastStep();
-            gradient = oldGradient;
-            parameters = oldParameters;
+            for (int i = 0; i < parameters.size(); ++i) {
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: See, above. Why change?

File: include/parpeoptimization/minibatchOptimization.h [hunk 5]
```diff
@@ -509,51 +509,94 @@ class MinibatchOptimizer {
...
+                gradient[i] = oldGradient[i];
+                parameters[i] = oldParameters[i];
+            }
 
+            // note the previous parameter step for debugging
+            for (int i = 0; i < parameters.size(); ++i)
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: Looks redundant with code above, move do separate function `difference`

File: include/parpeoptimization/minibatchOptimization.h [hunk 6]
```diff
@@ -509,51 +509,94 @@ class MinibatchOptimizer {
...
             // Do the next step
             learningRate = learningRateUpdater->getCurrentLearningRate();
             parameterUpdater->updateParameters(learningRate, iteration, gradient, parameters, lowerParameterBounds,
                                                upperParameterBounds);
 
             // Re-evaluate the cost function and hope for the best
             status = evaluate(f, parameters, datasets, cost, gradient, logger, reporter);
+            
+            // get Difference for debugging
+            for (int i = 0; i < parameters.size(); ++i)
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: Implement/use `difference`

File: include/parpeoptimization/minibatchOptimization.h [hunk 7]
```diff
@@ -601,7 +645,10 @@ class MinibatchOptimizer {
                    &logger, &reporter, this](double alpha) {
             
             /* Reset oldParameters and re-update with new step length */
-            oldParameters = parameters;
+            for (unsigned int i = 0; i < parameters.size(); ++i)
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: Why changed?

File: include/parpeoptimization/minibatchOptimization.h [hunk 8]
```diff
@@ -614,7 +661,17 @@ class MinibatchOptimizer {
...
+            logger->logmessage(LOGLVL_INFO, " Line-Search: Step was good right away...");
+            return;
+        }
+        
+        // Debugging output
+        std::stringstream line_ss;
+        std::stringstream parabola_ss;
+        std::vector<double> parDifference(parameters.size(), NAN);
+        for (unsigned int i = 0; i < parameters.size(); ++i)
+            parDifference[i] = parameters[i] - oldParameters[i];
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: ...

File: include/parpeoptimization/minibatchOptimization.h [hunk 9]
```diff
@@ -645,17 +705,35 @@ class MinibatchOptimizer {
...
          * Propose a new step based on a parabolic interpolation */
-        double newStepLength = 0.5 * dirGradient * std::pow(stepLength, 2.0) /
+        double newStepLength = -0.5 * dirGradient * std::pow(stepLength, 2.0) /
                 (cost1 - cost - dirGradient * stepLength);
         double cost2 = evalLineSearch(newStepLength);
         
+        // Debugging output
+        if (logger) {
+            line_ss << " Line-Search: two steps were done, results are: " << std::endl
+                << "   cost0: " << cost << " (steplength: " << getVectorNorm(parDifference)
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: What are cost0, cost1, cost2? Pls document

File: src/parpeoptimization/minibatchOptimization.cpp [hunk 10]
```diff
@@ -203,10 +203,12 @@ void ParameterUpdaterAdam::updateParameters(double learningRate,
     double tmpNumerator;
     double tmpDenominator;
 
-    oldGradientNormCache = gradientNormCache;
-    oldGradientCache = gradientCache;
-
     for (int i = 0; i < numParameters; ++i) {
```
Comment (dweindl) [inline src/parpeoptimization/minibatchOptimization.cpp]: ....

File: src/parpeoptimization/minibatchOptimization.cpp [hunk 11]
```diff
@@ -223,6 +225,11 @@ void ParameterUpdaterAdam::updateParameters(double learningRate,
 
 void ParameterUpdaterAdam::undoLastStep() {
     // The cached gradient norm needs to be restored, since the new one is probably NaN
+    for (int i = 0; i < gradientCache.size(); ++i) {
+        gradientNormCache[i] = oldGradientNormCache[i];
```
Comment (dweindl) [inline src/parpeoptimization/minibatchOptimization.cpp]: ...

File: include/parpeoptimization/minibatchOptimization.h (hunk 9 above)
Comment (paulstapor) [reply include/parpeoptimization/minibatchOptimization.h]: I wrote a longer comment on what these quantities are and what they do in lines 668 - 680. Is it okay like this or do you want a more formal documentation?

File: include/parpeoptimization/minibatchOptimization.h [hunk 12]
```diff
@@ -507,53 +514,98 @@ class MinibatchOptimizer {
...
             // Do the next step
             learningRate = learningRateUpdater->getCurrentLearningRate();
             parameterUpdater->updateParameters(learningRate, iteration, gradient, parameters, lowerParameterBounds,
                                                upperParameterBounds);
 
             // Re-evaluate the cost function and hope for the best
             status = evaluate(f, parameters, datasets, cost, gradient, logger, reporter);
+            
+            // get Difference for debugging
+            parDifference = getVectorDifference(parameters[i], oldParameters[i]);
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: See above. Remove `[i]`?

File: include/parpeoptimization/minibatchOptimization.h [hunk 13]
```diff
@@ -601,7 +654,9 @@ class MinibatchOptimizer {
                    &logger, &reporter, this](double alpha) {
             
             /* Reset oldParameters and re-update with new step length */
-            oldParameters = parameters;
+            std::copy(oldParameters.begin(), oldParameters.end(), parameters.begin())
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: Copies the other way as before.

File: include/parpeoptimization/minibatchOptimization.h [hunk 14]
```diff
@@ -610,11 +665,32 @@ class MinibatchOptimizer {
...
         
-        /* Do we check for a decreasing cost function? */
+        /* From here, we check for a decreasing cost function,
+         * otherwise, we already exited. 
+         * 
+         * The idea:
+         * cost0 is the previous cost before the parameter update,
+         * we want the update to be lower.
+         * 
+         * First compute a naive step as if there was now line-search -> get cost1
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]:
```suggestion
         * First compute a naive step as if there was no line-search -> get cost1
```

File: include/parpeoptimization/minibatchOptimization.h [hunk 15]
```diff
@@ -610,11 +665,32 @@ class MinibatchOptimizer {
...
-        if (cost1 <= cost) return;
+        if (cost1 <= cost) {
+            logger->logmessage(LOGLVL_DEBUG, " Line-Search: Step was good right away...");
+            return;
+        }
+        
+        // Debugging output
+        std::stringstream line_ss;
+        std::stringstream parabola_ss;
+        std::vector<double> parDifference = getVectorDifference(parameters[i], oldParameters[i]);
```
Comment (dweindl) [inline include/parpeoptimization/minibatchOptimization.h]: See above [i]

File: include/parpeoptimization/minibatchOptimization.h (hunk 9 above)
Comment (dweindl) [reply include/parpeoptimization/minibatchOptimization.h]: Works for me.
